From 55207ee582790668a468c1b42b551073132bbac3 Mon Sep 17 00:00:00 2001
From: Niyas Sait <niyas.sait@linaro.org>
Date: Tue, 7 Jan 2026 19:42:57 +0000
Subject: [PATCH] MSVC: Support spinlocks on AArch64

Backport of PostgreSQL commit a516b3f ("MSVC: Support building for
AArch64") for 17.9.  Co-authored-by: Greg Burd <greg@burd.me>
Co-authored-by: Dave Cramer <davecramer@gmail.com>

---
 src/include/storage/s_lock.h | 30 ++++++++++++++++++++++++++----
 1 file changed, 26 insertions(+), 4 deletions(-)

diff --git a/src/include/storage/s_lock.h b/src/include/storage/s_lock.h
index 2f73f9f..0c0ffd4 100644
--- a/src/include/storage/s_lock.h
+++ b/src/include/storage/s_lock.h
@@ -620,13 +620,24 @@ typedef LONG slock_t;
 
 #define SPIN_DELAY() spin_delay()
 
-/* If using Visual C++ on Win64, inline assembly is unavailable.
- * Use a _mm_pause intrinsic instead of rep nop.
- */
-#if defined(_WIN64)
+#ifdef _M_ARM64
+static __forceinline void
+spin_delay(void)
+{
+	/*
+	 * Research indicates ISB is better than __yield() on AArch64. See
+	 * https://postgr.es/m/1c2a29b8-5b1e-44f7-a871-71ec5fefc120%40app.fastmail.com.
+	 */
+	__isb(_ARM64_BARRIER_SY);
+}
+#elif defined(_WIN64)
 static __forceinline void
 spin_delay(void)
 {
+	/*
+	 * If using Visual C++ on Win64, inline assembly is unavailable.
+	 * Use a _mm_pause intrinsic instead of rep nop.
+	 */
 	_mm_pause();
 }
 #else
@@ -639,6 +650,15 @@ spin_delay(void)
 #endif
 
 #include <intrin.h>
+
+#ifdef _M_ARM64
+
+/* _ReadWriteBarrier() is insufficient on non-TSO architectures. */
+#pragma intrinsic(_InterlockedExchange)
+#define S_UNLOCK(lock) _InterlockedExchange(lock, 0)
+
+#else
+
 #pragma intrinsic(_ReadWriteBarrier)
 
 #define S_UNLOCK(lock)	\
@@ -646,6 +666,8 @@ spin_delay(void)
 
 #endif
 
+#endif
+
 
 #endif	/* !defined(HAS_TEST_AND_SET) */
 
--
2.30.1
